OFEP-sdk-wait-provider-ready.md

```diff
@@ -0,0 +1,58 @@
+## OFEP-sdk-wait-provider-ready
+
+## State: DRAFTING
+Implement a mechanism to wait for the provider to be in a ready state.
+
+## Background
+
+Provider can now define a `initialize` function called by the SDK when Openfeature users call the function `setProvider(...)`. 
+The `initialize` function is called asynchronously and it is possible to get the information that the provider is ready while awaiting an initial ready event.
+
+In some languages such as `javascript` waiting for the event could be one line of code, but on other languages it could need way more boiler plate code to do that.
+```javascript
+Openfeature.setProvider(my-provider)
+const client = Openfeature.getClient()
+await new Promise((resolve) => client.on(‘ready’, resolve))
+```
+
+```java
+OpenFeatureAPI.getInstance().setProvider("test", g);
+Client cli = OpenFeatureAPI.getInstance().getClient("test");
+CompletableFuture<EventDetails> completableFuture = new CompletableFuture<>();
+OpenFeatureAPI.getInstance().onProviderReady(new Consumer<EventDetails>() {
+    @Override
+    public void accept(EventDetails eventDetails) {
+        completableFuture.complete(eventDetails);
+    }
+});
+completableFuture.get();
+```
+
+## Proposal
+It would be great to have a way in the SDKs to wait for the provider to be ready.
+
+
+## Alternatives
+
+### Alternative 1: make `setProvider` synchronous
+Probably not the most ideal solution but we could change the behavior of `setProvider` to wait for the initialize function to throw or return.  
+*It is a breaking change from the actual implementation.*
+
+### Alternative 2: Add a waiting function in the SDKs to wait for the provider to be ready
```

Comment:
I think this is reasonable. It provides a good deal of flexibility. I think, depending on language, you still end up also adding a timeout. Waiting forever isn't a great idea, and similar to the initial problem not all languages provide the ease of handling the timeout. (In JS you just race the promise with a time promise.)

`waitReady(timeout)`.

I do wonder if there is any value in difference here between core spec and stuff you can already build with the spec? It is inconvenient in java, but not impossible. Is there some other level we should have for things that can be implemented as a layer? I suppose if it is just an optional part of the spec that would be sufficient.

Reply:
> I think this is reasonable. It provides a good deal of flexibility. I think, depending on language, you still end up also adding a timeout. Waiting forever isn't a great idea, and similar to the initial problem not all languages provide the ease of handling the timeout. (In JS you just race the promise with a time promise.)

Agreed 100%.

> I do wonder if there is any value in difference here between core spec and stuff you can already build with the spec? It is inconvenient in java, but not impossible. Is there some other level we should have for things that can be implemented as a layer? I suppose if it is just an optional part of the spec that would be sufficient.

I've been considering something like this as well, but I think an optional requirement (MAY or SHOULD) would be sufficient for this.

MAY or SHOULD, in combination with the nature of the requirement, strongly imply this would be a convenience method, which I believe is basically the case here.